File: Java/Hussain Sir Java/Fundamentals/Generics/Basics/IterationTest.java
import java.util.Iterator;

class SimpleStack<V> implements Iterable<V>{
	
	private Node top;

	class Node{
		V value;
		Node below;

		Node(V val){
			value = val;
			below = top;
		}
	}

	public void push(V val){
		top = new Node(val);
	}

	public V pop(){
		Node node = top;
		top = top.below;
		return node.value;
	}

	public boolean empty(){
		return top == null;
	}

	public Iterator<V> iterator(){
		return new Iterator<V>(){
			Node current = top;

			public boolean hasNext(){
				return current != null;
			}

			public V next(){
				V val = current.value;
				current = current.below;
				return val;
			}

		};
	}
}

class IterationTest{

	public static void main(String[] args){
		int[] numbers = {1, 2, 3, 4, 5, 6};
		System.out.println("All integers in array");
		for(int n : numbers)
			System.out.println(n);
		SimpleStack<String> days = new SimpleStack<>();
		days.push("Monday");
		days.push("Tuesday");
		days.push("Wednesday");
		days.push("Thursday");
		days.push("Friday");
		System.out.println("All strings on stack");
		for(String d : days)
			System.out.println(d);
	}
}

File: Fundamentals/Generics/Basics/iterationtest.cs
using System;
using System.Collections.Generic;

class SimpleStack<V>
{
	class Node
	{
		internal V value;
		internal Node below;

		internal Node(V val, Node node)
		{
			value = val;
			below = node;
		}
	}

	private Node top;

	public void Push(V value)
	{
		top = new Node(value, top);
	}

	public V Pop()
	{
		Node node = top;
		top = top.below;
		return node.value;
	}

	public bool Empty()
	{
		return top == null;
	}
	
	public IEnumerator<V> GetEnumerator()
	{
		for(Node n = top; n != null; n = n.below)
			yield return n.value;		
	}
}

static class Program
{
	public static void Main()
	{
		int[] a = {1, 4, 9, 16, 25};
		Console.WriteLine("All integers in array");
		foreach(int i in a)
			Console.WriteLine(i);

		SimpleStack<string> b = new SimpleStack<string>();
		b.Push("June");
		b.Push("May");
		b.Push("April");
		b.Push("March");
		Console.WriteLine("All strings on stack");
		foreach(string s in b)
			Console.WriteLine(s);
	}
}
